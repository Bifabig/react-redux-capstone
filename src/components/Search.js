import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink, useLocation } from 'react-router-dom';
import { getSearchResult } from '../redux/symbols/symbolsSlice';
import styles from '../styles/Search.module.css';

const Search = () => {
  const location = useLocation();
  const dispatch = useDispatch();
  const { searchSymbols, isLoading, error } = useSelector(
    (state) => state.symbols,
  );
  useEffect(() => {
    if (location.state.searchVal !== '') {
      dispatch(getSearchResult(location.state.searchVal));
    }
    location.state.searchVal = '';
  }, [dispatch, location.state, location.state.searchVal]);

  if (error) return <h2>Something went wrong</h2>;

  return isLoading ? (<h2>Loading...</h2>) : (
    <div className={styles.container}>
      <NavLink to="/" className={styles.back}>
        <i className="bi bi-arrow-left-circle" />
      </NavLink>
      <h2>Search Results</h2>
      { searchSymbols.length !== 0 ? searchSymbols.map((symbol) => (
        <div className={styles.listItems} key={symbol}>
          <NavLink to="/balanceSheets" state={{ symbol: { symbol } }}>
            <li>

              {symbol}
              <i className="bi bi-arrow-right-circle" />
            </li>
          </NavLink>

        </div>
      )) : <h4 className={styles.container}> Your search item doesn&apos;t exist</h4>}

    </div>
  );
};

export default Search;
