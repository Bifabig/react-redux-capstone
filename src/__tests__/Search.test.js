import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import Search from '../components/Search';

import {
  getSearchResult,
} from '../redux/symbols/symbolsSlice';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/symbols/symbolsSlice', () => ({
  getSearchResult: jest.fn(),
}));

describe('Search', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockReturnValue({
      searchSymbols: [],
      allSymbols: [],
      symbols: [],
      isLoading: false,
      error: undefined,
      hasMore: true,
    });
  });

  it('should dispatch getSearchResult when symbols array is empty', () => {
    useSelector.mockReturnValue({
      searchSymbols: [],
      symbols: [],
      error: null,
      hasMore: true,
    });

    const { container } = render(<MemoryRouter initialEntries={['/search', { state: { searchVal: 'hi', isSubmitted: true } }]}><Search /></MemoryRouter>);

    expect(getSearchResult).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
