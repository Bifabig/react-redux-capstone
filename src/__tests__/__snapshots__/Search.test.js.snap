// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should dispatch getSearchResult when symbols array is empty 1`] = `
<div>
  <div
    class="container"
  >
    <a
      aria-current="page"
      class="back active"
      href="/"
    >
      <i
        class="bi bi-arrow-left-circle"
      />
    </a>
    <h2>
      Search Results
    </h2>
    <h4
      class="container"
    >
       Your search item doesn't exist
    </h4>
  </div>
</div>
`;
